app/actions/websocket/roles.ts

```diff
@@ -0,0 +1,111 @@
+    const database = DatabaseManager.serverDatabases[serverUrl];
+    if (!database) {
+        return;
+    }
+
+    // only update Role records that exist in the Role Table
+    const role = JSON.parse(msg.data.role);
+    const dbRole = await queryRoleById(database.database, role.id);
+    if (!dbRole) {
+        return;
+    }
+
+    await database.database.write(async () => {
+        await dbRole.update((roleRecord: RoleModel) => {
+            roleRecord.permissions = role.permissions;
+        });
+    });
+}
+
+export async function handleUserRoleUpdatedEvent(serverUrl: string, msg: WebSocketMessage): Promise<void> {
+    const database = DatabaseManager.serverDatabases[serverUrl];
+    if (!database) {
+        return;
+    }
+
+    const currentUserId = await queryCurrentUserId(database.database);
+    if (currentUserId !== msg.data.user_id) {
+        return;
+    }
+
+    const modelPromises: Array<Promise<Model[]>> = [];
+
+    // update Role Table if needed
+    const newRoles = await fetchRolesIfNeeded(serverUrl, Array.from(msg.data.roles), true);
+    const preparedRoleModels = getPreparedRoleModels(database, newRoles);
+    modelPromises.push(preparedRoleModels);
+
+    // update User Table record
+    const meData = await fetchMe(serverUrl, true);
+    const userModels = prepareUsers(database.operator, [meData.user!]);
+    if (userModels) {
+        modelPromises.push(userModels);
+    }
+
+    const models = await Promise.all(modelPromises);
+    const flattenedModels = models.flat() as Model[];
+    if (flattenedModels?.length > 0) {
+        await database.operator.batchRecords(flattenedModels);
+    }
+}
+
+export async function handleMemberRoleUpdatedEvent(serverUrl: string, msg: WebSocketMessage): Promise<void> {
+    const database = DatabaseManager.serverDatabases[serverUrl];
+    if (!database) {
+        return;
+    }
+
+    const currentUserId = await queryCurrentUserId(database.database);
+    const member = JSON.parse(msg.data.member);
+    if (currentUserId !== member.user_id) {
+        return;
+    }
+
+    const modelPromises: Array<Promise<Model[]>> = [];
+
+    // update Role Table if needed
+    const newRoles = await fetchRolesIfNeeded(serverUrl, Array.from(member.roles), true);
+    const preparedRoleModels = getPreparedRoleModels(database, newRoles);
+    modelPromises.push(preparedRoleModels);
+
+    // update MyTeam Table
+    const preparedMyTeam = prepareMyTeams(database.operator, [], [member]);
+    if (preparedMyTeam) {
+        modelPromises.push(preparedMyTeam[2]);
+    }
+
+    const models = await Promise.all(modelPromises);
+    const flattenedModels = models.flat() as Model[];
+    if (flattenedModels?.length > 0) {
+        await database.operator.batchRecords(flattenedModels);
+    }
+}
+
+async function getPreparedRoleModels(database: ServerDatabase, roles: RolesRequest): Promise<Model[]> {
+    if (!(typeof roles.roles === 'string' && roles.roles === 'null')) {
```

Comment:
When does this happen? Seems like something is wrong somewhere for this to be handled like this.

Reply:
fixed and removed completely.